src/reconciler/hotReplacementRender.js

```diff
@@ -348,6 +348,21 @@ const hotReplacementRender = (instance, stack) => {
   }
 }
 
+export const hotComponentCompare = (oldType, newType) => {
+  if (oldType === newType) {
+    return true
+  }
+
+  if (isSwappable(newType, oldType)) {
```

Comment:
This is the actual change.
If type == type - return.
Else - compare them. Both types are "wrapped".
If types are replaceable
 - rollback proxy1
 - set alias oldProxy->newProxy
 - set alias newProxy(usedId) -> oldProxy (may be obsolete)

return true.

Result it 💯 full cream hot-loader.